loglist/logfilter_test.go

```diff
@@ -0,0 +1,119 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package loglist
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/certificate-transparency-go/testdata"
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509util"
+
+	"github.com/mohae/deepcopy"
+)
+
+func subLogList(t *testing.T, logURLs map[string]bool) LogList {
+	t.Helper()
+	var ll LogList
+	ll.Operators = deepcopy.Copy(sampleLogList.Operators).([]Operator)
+	for _, l := range sampleLogList.Logs {
+		if logURLs[l.URL] {
+			ll.Logs = append(ll.Logs, l)
+		}
+	}
+	return ll
+}
+
+func TestActiveLogs(t *testing.T) {
+	tests := []struct {
+		name string
+		in   LogList
+		want LogList
+	}{
+		{
+			name: "Sample",
+			in:   sampleLogList,
+			want: subLogList(t, map[string]bool{"ct.googleapis.com/aviator/": true, "ct.googleapis.com/icarus/": true, "ct.googleapis.com/rocketeer/": true, "ct.googleapis.com/racketeer/": true}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got := test.in.ActiveLogs()
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("Extracting active logs out of %v. \nExpected %v \nwhile got %v.", test.in, test.want, got)
+			}
+		})
+	}
+}
+
+func certChain(t *testing.T) []*x509.Certificate {
+	t.Helper()
+	cert, _ := x509util.CertificateFromPEM([]byte(testdata.TestPreCertPEM))
+	caCert, _ := x509util.CertificateFromPEM([]byte(testdata.CACertPEM))
+	var chain []*x509.Certificate
+	chain = append(chain, cert)
+	chain = append(chain, caCert)
+	return chain
+}
+
+func singleCert(t *testing.T) []*x509.Certificate {
+	t.Helper()
+	cert, _ := x509util.CertificateFromPEM([]byte(testdata.TestPreCertPEM))
+	var chain []*x509.Certificate
+	chain = append(chain, cert)
+	return chain
+}
+
+func artificialRoots(t *testing.T) LogRoots {
+	m := make(LogRoots)
+	caCert, _ := x509util.CertificateFromPEM([]byte(testdata.CACertPEM))
+	m["log.bob.io"] = []*x509.Certificate{caCert}
+	return m
+}
+
+func TestCompatible(t *testing.T) {
+	tests := []struct {
+		name  string
+		in    LogList
+		chain []*x509.Certificate
+		roots LogRoots
+		want  LogList
+	}{
+		{
+			name:  "RootedChain",
+			in:    sampleLogList,
+			chain: certChain(t),
+			roots: artificialRoots(t),
+			want:  subLogList(t, map[string]bool{"log.bob.io": true}),
+		},
+		{
+			name:  "UnRootedChain",
+			in:    sampleLogList,
+			chain: singleCert(t),
+			roots: artificialRoots(t),
+			want:  subLogList(t, map[string]bool{}),
+		},
```

Comment:
Please add test cases for cert chains containing an intermediate CA cert that:
a) is trusted by a log
b) is not trusted by any log

Reply:
As noted: the func is not expected to process incomplete chains.

Reply:
It wouldn't be an incomplete chain. It'd actually be a chain with more certificates in it than necessary.

Reply:
Added testcase.

Reply:
Thanks. It'd be nice to have tests for longer chains too (e.g. 3 or 4 entries).